RecoLocalTracker/Phase2TrackerRecHits/test/RecHitsValidation.cc

```diff
@@ -264,13 +487,37 @@ std::map< unsigned int, RecHitHistos >::iterator Phase2TrackerRecHitsValidation:
     return insertedIt.first;
 }
 
+
 unsigned int Phase2TrackerRecHitsValidation::getLayerNumber(const DetId& detid, const TrackerTopology* topo) {
     if (detid.det() == DetId::Tracker) {
-        if (detid.subdetId() == PixelSubdetector::PixelBarrel) return (topo->pxbLayer(detid));
-        else if (detid.subdetId() == PixelSubdetector::PixelEndcap) return (100 * topo->pxfSide(detid) + topo->pxfDisk(detid));
-        else return 999;
+        if (detid.subdetId() == PixelSubdetector::PixelBarrel) {
+          return (topo->pxbLayer(detid));
+        } else if (detid.subdetId() == PixelSubdetector::PixelEndcap) {
+          return (100 * topo->side(detid) + topo->pxfDisk(detid));
+        } else if (detid.subdetId() == StripSubdetector::TIB) {
+          return (topo->layer(detid));
+        } else if (detid.subdetId() == StripSubdetector::TOB) {
+          return (topo->layer(detid));
+        } else if (detid.subdetId() == StripSubdetector::TID) {
+          return (100 + topo->tidRing(detid));
+        } else if (detid.subdetId() == StripSubdetector::TEC) {
+          return (100 + topo->tecRing(detid));
+        } else {
+	  return 999;
+	}
```

Comment:
I agree that those methods in TrackerTopology can be source of confusion and pain in future. Whatever depends on the geometry has to be in TrackerGeometry and not in TrackerTopology.
I also agree that TrackerTopology::layer() has to be used in this case. 
I also suggest to use the enumerators like StripSubdetector::TIB as less as possible. They do not contain anymore any information about the actual geometry of the subdetector: they are just integers from 1 to 6. They have to be used _only_ to decide which TrackerTopology method has to be called (as it is correctly done in this case but it is not needed because of the generic layer() method). If you have to know what is the subdetector you should use the methods in TrackerGeometry here: https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h#L81